UM/Settings/MachineManager.py

```diff
@@ -67,6 +71,46 @@ def getAllMachineVariants(self, machine_id):
 
         return variants
 
+    def getAllMachineMaterials(self, instance_id):
+        generic_materials = []
+        machine_materials = []
+
+        machine = self.findMachineInstance(instance_id)
+        if not machine:
+            return machine_materials
+            
+        machine_type = machine.getMachineDefinition().getProfilesMachineId()
+        machine_variant = machine.getMachineDefinition().getVariantName()
+
+        for profile in self._profiles:
+            profile_type = profile.getType()
+            #Filter out "partial" profiles
+            if profile_type == "material":
+                generic_materials.append(profile.getName())
+                continue
+
+            material = profile.getMaterialName()
+            if not material or material in machine_materials:
+                continue
+
+            profile_machine_type = profile.getMachineTypeId()
+            profile_machine_variant = profile.getMachineVariantName()
+            profile_machine_instance = profile.getMachineInstanceName()
+
+            if (profile_machine_instance and profile_machine_instance == instance_id) or \
+                    (profile_machine_variant and profile_machine_variant == machine_variant and profile_machine_type == machine_type) or \
+                    (profile_machine_type == machine_type):
+                machine_materials.append(material)
+
+        if len(machine_materials) > 0:
+            return machine_materials
+        elif machine.getMachineDefinition().getSetting("machine_gcode_flavor").getDefaultValue() == "UltiGCode":
+            #UltiGCode printers don't use generic materials in Cura
+            return []
```

Comment:
No, because there are UltiGCode printers that will have material profiles (UM2+, UM3) (defeating the whole purpose of ultigcode, I know). I'll add more/better documentation.